### DISCUSSION TITLE
How to pass parameter from main lambda handler to event handlerroute

### DISCUSSION THREAD

**Discussion by MaartenUreel** [aws-powertools/powertools-lambda-python#1546] [Q&A] [answered]
Assume a main lambda method, where we construct some variable:

```python
@logger.inject_lambda_context()
@event_source(data_class=LambdaFunctionUrlEvent)
def lambda_handler(event: LambdaFunctionUrlEvent, context: LambdaContext) -> dict:
    # Parse & construct some parameters
    parameters = dict(parse_qsl(event.decoded_body))

    # Create a CallContext object
    call_context = CallContext(db, parameters)

    return app.resolve(event, context)
```

What is a good way to pass this `call_context` variable to a route handler, e.g.:

```python
@app.post("/dial_call_status")
def dial_call_status():
    # How to access call_context here?

```

**Comment by heitorlessa** [comment]
Hey @MaartenUreel, we don't provide a dependency injection (DI) feature in Event Handler at the moment, so you'd have to handle `call_context` in the context of your route -- Using your example, it'd look like:

```python
@app.post("/dial_call_status")
def dial_call_status():
    parameters = dict(parse_qsl(app.current_event.decoded_body))

    # Create a CallContext object
    call_context = CallContext(db, parameters)
``` 

If your need is more advanced, we'd love to [discuss how a DI would look like in a RFC format](https://github.com/awslabs/aws-lambda-powertools-python/issues/new?assignees=&labels=RFC%2Ctriage&template=rfc.yml&title=RFC%3A+TITLE) for Event Handler. As DI has some cognitive load, I'd be looking at **(1)** UX of providing/depending on X, **(2)** whether any Callable suffice, **(3)** whether we should call it only once if a dependency is used multiple times (e.g., cache=True), and **(4)** whether we want to limit the depth of a graph (e.g., sub-sub-sub-dependencies).

Thanks a lot!

**Comment by MaartenUreel** [comment]
OK; I hadded it to `context.__dict__["call_context"]` and was able to retrieve it in the handler with `app.lambda_context.__dict__["call_context"]` then; it's a bit hacky but it will do for now.

**Comment by MaartenUreel** [comment, marked as answer]
[I opened up an RFC](https://github.com/awslabs/aws-lambda-powertools-python/issues/1547) anyway because I find I have too much boilerplate code now in every request.
I hope I did it according to standards ;-)

**Reply by heitorlessa** [reply]
Thanks a lot, marking this as the answer and will comment in the RFC today ;)
